text/0000-kafka-topics-per-tenant.md

```diff
@@ -0,0 +1,100 @@
+- **Cost**: Kafka's use in FOLIO is still limited, more FOLIO modules will continue to employ event driven techniques that will cause an explosion of partitions. Adding more brokers to service the load compounded by partitions will mean increased costs.
+
+## Out of Scope
+- Enabling tenant data security within a Kafka installation. Tenant data security meaning "How do we prevent a tenant's data to be processed in place of another?", "How can we prevent a tenant module from seeing data from other tenants?"
+- Moving forward with the [Temporary Kafka Security Solution](https://wiki.folio.org/x/YYVFAw).
+- Countering module multi-versioning i.e. more than one version of a module is installed in FOLIO with tenants able to target specific versions.
+- FOLIO modules that do not use folio-kafka-wrapper to interact with Kafka. These modules should be migrated to folio-kafka-wrapper to ease partition counts.
+
+## Rationale and Alternatives
+The main driver for this RFC is to reduce partition count within reasonable means. Ceasing to create Kafka topics for each tenant will get us to a more reasonable count. Further feature development on FOLIO can increase the partition count, this is expected.
+
+- **Merging All Topics Into A Single Topic**:This would be a maintenance nightmare. Messages could be consumed in error because so many consumer groups will be present and would rely on application code to filter out irrelevant messages for any particular module.
+- **Have A Single Topics For Each Module Area**: Module areas being Inventory, Acquisitions, Circulation etc. This shares the same pain as the previous point but to a lesser extent.
+
+
+## Detailed Explanation/Design
+### Consuming
+In [folio-kafka-wrapper](https://github.com/folio-org/folio-kafka-wrapper), a sample of a subscription pattern is defined below
+
+>folio\.Default\.\w{1,}\.DI_COMPLETED
+
+where `folio` is the cluster name, `Default` is the namespace then a wildcard to capture tenant ids followed by the event type(`DI_COMPLETED`) that the topic is supposed to contain.
+Subscriptions are created using this interface:
+```
+public static SubscriptionDefinition createSubscriptionDefinition(String env, String nameSpace, String eventType)
+```
+Modifications will be made to mark the method above as deprecated and another method created to remove the wildcard which will allow the subscription pattern to look like below
+>folio\.Default\.DI_COMPLETED
+
+Safeguards will be added to ensure that a message has a tenant id as one of its headers. It is imperative that every message has a owner.
```

Comment:
when folio-kafka-wrapper is used, logic can be added to ensure that a kafka message is enriched with the tenant id before the message is sent. In addition, messages consumed can ensure the presence of the tenant id and discard messages that don't have a tenant id. Interfaces can also be designed to communicate to a developer that the message is being consumed within the context of a tenant id - for example returning the kafka message and tenant id to downstream code as a pair/collection.

Reply:
Thank you for explaining the kinds of safe guards you are considering.

> In addition, messages consumed can ensure the presence of the tenant id and discard messages that don't have a tenant id. 

Are these going to be reported in any way?

My concern is that this could make diagnosing issues like failures to synchronise data between modules hard to do.

Reply:
That is a valid concern. I have listed the safeguards in the RFC.
> Are these going to be reported in any way?

The message can be logged as a warning or we can push for a dead queue approach where messages that could not be processed are placed into another topic for further review.

Reply:
Cool. 

I think a mixed tenant topic makes it more likely for those circumstances to happen, so ideally would want something in place to help diagnose it when it happens in production.

I'll resolve this comment.